src/components/Firestore/Requests/RequestPath/index.tsx

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../index.scss';
+
+import { IconButton } from '@rmwc/icon-button';
+import { Tooltip } from '@rmwc/tooltip';
+import React, { useEffect, useRef, useState } from 'react';
+
+// Copies path without spaces to clipboard
+// and triggers copy notification (SnackBar)
+function copyPathToClipboard(
+  resourcePath: string,
+  setShowCopyNotification: (value: boolean) => void
+) {
+  navigator.clipboard
+    .writeText(resourcePath.replace(/\s/g, ''))
+    .then(() => setShowCopyNotification(true))
+    .catch(() => setShowCopyNotification(false));
+}
+
+// Function that truncates the request path from the left side of the string.
+// It first calculates the available width of the pathHtmlElement, and how many
+// characters fit inside the width. Based on that, it calculates a new truncated string.
+// For optimization purposes, the truncation is always calculated, but the HTMLElement
+// is only updated when the new truncated string is different than the HTMLElement's textContent.
+function truncateRequestPathFromLeft(
+  pathTextElement: React.RefObject<HTMLDivElement>,
```

Comment:
I'm not too excited about having properties only used for testing. Here's one suggestion to avoid this:

1. Create a new (exported for testing) function that accepts `pathHtmlElement` directly as an `HTMLDivElement`.
2. Have this function pull out the `.current` value and just call the new function
3. Add some tests that directly call this new function. It won't go through the DOM, but that's okay in this case
    1. Now, you can pass in a plain objects with `textContent` and `offsetWidth` properties. (You'll need to cast them with `as unknown as HTMLDivElement` or `as unknown as HTMLButtonElement`)
    2. After the function call, you can check the text was truncated correctly by examining the `textContent` field on this mock object

Your test will no longer directly check the interaction with the DOM, but I think that's worth it in this case to avoid the hack

Reply:
I added one solution from what we dicussed yesterday in gChat:

- Now the `offsetWidth` properties of the **request-path** and the **copy-icon-button** are mocked inside the test file, therefore, the UI is not modified at all (no need to inject test-props).
- With this approach, the component can be rendered in the DOM.

Let me know what you think.